src/nupic/support/group_by.py

```diff
@@ -0,0 +1,130 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+""" A utility function wrapper based on groupby in itertools. Allows to
+    walk across n sorted lists with respect to their key functions
+    and yields a tuple of n lists of the members of the next *smallest*
+    group.
+
+    @param args (list) a list of arguments alternating between sorted lists and
+                       their respective key functions. The lists should be
+                       sorted with respect to their key function.
+
+    @return (tuple) a n + 1 dimensional tuple, where the first element is the
+                    key of the group and the other n entries are lists of
+                    objects that are a member of the current group that is being
+                    iterated over in the nth list passed in. Note that this
+                    is a generator and a n+1 dimensional tuple is yielded for
+                    every group.
+                    
+    Notes: Read up on groupby here:
+           https://docs.python.org/dev/library/itertools.html#itertools.groupby
+
+"""
+
+class GroupByGenerator(object):
+  
+  __slots__ = ['iterable', 'length', 'index']
+
+  def __init__(self, iterable, length):
+    self.iterable = iterable
+    self.length = length
+    self.index = 0
+
+
+  def __iter__(self):
+    for i in xrange(self.length):
+      self.index = i
+      yield self.iterable[i]
+
+
+  def __len__(self):
+    return self.length - self.index
+
+
+def groupby(lis, fn):
+  length = 1
+  begining = 0
+
+  for i in xrange(len(lis)):
+    val = fn(lis[i])
+
+    if i == 0:
+      key = val
+      continue
+
+    if val != key:
+      yield (key, GroupByGenerator(lis[begining : i], length))
```

Comment:
This `lis[begining : i]` is still creating a new list. Can we just pass in `lis`, `beginning`, and `length` to the GroupByGenerator?